pkg/sql/user.go

```diff
@@ -288,92 +291,98 @@ func retrieveAuthInfo(
 	// we should always look up the latest data.
 	const getHashedPassword = `SELECT "hashedPassword" FROM system.public.users ` +
 		`WHERE username=$1`
-	ie := f.Executor()
-	values, err := ie.QueryRowEx(
-		ctx, "get-hashed-pwd", nil, /* txn */
-		sessiondata.NodeUserSessionDataOverride,
-		getHashedPassword, user)
-
-	if err != nil {
-		return aInfo, errors.Wrapf(err, "error looking up user %s", user)
-	}
-	var hashedPassword []byte
-	if values != nil {
-		aInfo.UserExists = true
-		if v := values[0]; v != tree.DNull {
-			hashedPassword = []byte(*(v.(*tree.DBytes)))
+	err := f.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
+		if err := txn.Descriptors().MaybeSetReplicationSafeTS(ctx, txn.KV()); err != nil {
+			return err
+		}
+		values, err := txn.QueryRowEx(
+			ctx, "get-hashed-pwd", txn.KV(), /* txn */
+			sessiondata.NodeUserSessionDataOverride,
+			getHashedPassword, user)
+		if err != nil {
+			return err
 		}
-	}
-	aInfo.HashedPassword = password.LoadPasswordHash(ctx, hashedPassword)
 
-	if !aInfo.UserExists {
-		return aInfo, nil
-	}
+		var hashedPassword []byte
+		if values != nil {
+			aInfo.UserExists = true
+			if v := values[0]; v != tree.DNull {
+				hashedPassword = []byte(*(v.(*tree.DBytes)))
+			}
+		}
 
-	// None of the rest of the role options are relevant for root.
-	if user.IsRootUser() {
-		return aInfo, nil
-	}
+		aInfo.HashedPassword = password.LoadPasswordHash(ctx, hashedPassword)
 
-	// Use fully qualified table name to avoid looking up "".system.role_options.
-	const getLoginDependencies = `SELECT option, value FROM system.public.role_options ` +
-		`WHERE username=$1 AND option IN ('NOLOGIN', 'VALID UNTIL', 'NOSQLLOGIN', 'REPLICATION', 'SUBJECT')`
+		if !aInfo.UserExists {
+			return nil
+		}
 
-	roleOptsIt, err := ie.QueryIteratorEx(
-		ctx, "get-login-dependencies", nil, /* txn */
-		sessiondata.NodeUserSessionDataOverride,
-		getLoginDependencies,
-		user,
-	)
+		// None of the rest of the role options are relevant for root.
+		if user.IsRootUser() {
+			return nil
+		}
 
-	if err != nil {
-		return aInfo, errors.Wrapf(err, "error looking up user %s", user)
-	}
-	// We have to make sure to close the iterator since we might return from
-	// the for loop early (before Next() returns false).
-	defer func() { retErr = errors.CombineErrors(retErr, roleOptsIt.Close()) }()
+		// Use fully qualified table name to avoid looking up "".system.role_options.
+		const getLoginDependencies = `SELECT option, value FROM system.public.role_options ` +
+			`WHERE username=$1 AND option IN ('NOLOGIN', 'VALID UNTIL', 'NOSQLLOGIN', 'REPLICATION', 'SUBJECT')`
 
-	// To support users created before 20.1, allow all USERS/ROLES to login
-	// if NOLOGIN is not found.
-	aInfo.CanLoginSQLRoleOpt = true
-	aInfo.CanLoginDBConsoleRoleOpt = true
-	var ok bool
+		roleOptsIt, err := txn.QueryIteratorEx(
+			ctx, "get-login-dependencies", txn.KV(), /* txn */
```

Comment:
you're fixing this in this PR, but it seems a bit odd that we previously grabbed the login dependencies in a separate txn than the hashed password.

Reply:
Yeah,  I think it was just a subtle miss. I don't think it was because the txn complexity would be too high here, since their both scanning for one row.
<!-- Reviewable comment -OCYpvAa92Ksi29Bobw_:b-j3cumm -->
<!-- Sent from Reviewable.io -->

Reply:
Done.

Also added:

// We are going to use a single txn for resolving the user information,  
// and role\_options for the user.
<!-- Reviewable comment -OCfS9-aCPjnAVXJlxEM:b-bisfbo -->
<!-- Sent from Reviewable.io -->